### PR TITLE
SOLR-16292 : NVector for alternative great-circle distance calculations

### PR DESCRIPTION
**Pull request by danrosher**
https://issues.apache.org/jira/browse/SOLR-16292

# Description

[N-Vector](https://en.wikipedia.org/wiki/N-vector#Example:_Great_circle_distance) is a three-parameter representation that can be used to calculate the great-circle distance (assuming a spherical Earth).

It uses FastInvTrig, which for small distances it compares well with Math.acos. NVector is a faster way of calculating the great-circle distance than Haversine.

# Solution

Store N-Vectors in solr index via CoordinateFieldType with 3 values for the nvector into single value double subfields, use java Math class for indexing these
Use an maclaurin approximation for acos for calculating great-circle distance at query-time via a function query

# Tests

Tests for the FastInvTrigTest impl to compare it's acos with Math.acos.
Tests for NVectorUtil, NVectorDist , latLongToNVector and NVectorToLatLong
Tests for indexing N-Vectors and calculating the great-circle distance via the function query.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/schema/NVector.java [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+            //Parse out a comma separated list of values, as in: 73.5,89.2,7773.4
+            for (; i < dimension; i++) {
+                while (start < end && externalVal.charAt(start) == ' ') start++;
+                while (end > start && externalVal.charAt(end - 1) == ' ') end--;
+                if (start == end) {
+                    break;
+                }
+                out[i] = externalVal.substring(start, end);
+                start = idx + 1;
+                end = externalVal.indexOf(',', start);
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/schema/NVector.java:109]: Numbers will always be stored as decimals with `.` and separator as `,`? Need to make sure this doesn't mess up in other locales that use comma as the decimal separator.

File: solr/core/src/java/org/apache/solr/search/function/distance/NVector.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        };
+    }
+
+    protected String name() {
+        return "nvector";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this.getClass() != o.getClass()) return false;
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/search/function/distance/NVector.java:83]: `o` could be null

File: solr/core/src/java/org/apache/solr/search/function/distance/NVector.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+import java.io.IOException;
+import java.util.Map;
+
+import static org.apache.solr.util.NVectorUtil.NVectorDist;
+
+public class NVector extends ValueSource {
+
+    private final MultiValueSource p1;
+    private final MultiValueSource p2;
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/search/function/distance/NVector.java:36]: I'm unclear what these are - I don't think they're lat/long. Are they some offset and phase? Two separate vectors? I don't think p1 and p2 are descriptive enough for non-domain experts (hi! that's me!)

File: solr/core/src/test-files/solr/collection1/conf/schema-nvector.xml [hunk 4]
```diff
@@ -0,0 +1,930 @@
...
+<!-- The Solr schema file. This file should be named "schema.xml" and
+     should be located where the classloader for the Solr webapp can find it.
+
+     This schema is used for testing, and as such has everything and the
+     kitchen sink thrown in. See example/solr/conf/schema.xml for a
+     more concise example.
+
+  -->
+
+<schema name="test" version="1.0">
```
Comment (madrob) [inline solr/core/src/test-files/solr/collection1/conf/schema-nvector.xml:28]: can we minimize this schema to only the fields that it needs instead of the full 1000 lines of everything?

File: solr/core/src/test/org/apache/solr/util/FastInvTrigTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+        double latitude = deg2rad(Math.toDegrees(Math.acos(u * 2 - 1)) - 90);
+        double longitude = deg2rad(360 * v - 180);
+        return new double[]{latitude, longitude};
+    }
+
+    @Test
+    public void acos() {
+        for (double i = -1; i <= 1; i = i + 0.00001) {
+            assertTrue(FastInvTrig.acos(i) - Math.acos(i) <= EPSILON);
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/util/FastInvTrigTest.java:60]: nit: use assertEquals to get better error messages

File: solr/core/src/test/org/apache/solr/util/FastInvTrigTest.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.util.Random;
+
+import static org.apache.solr.util.NVectorUtil.EARTH_RADIUS;
+
+public class FastInvTrigTest extends SolrTestCase {
+
+    final static int num_points = 100000;
+    final static double EPSILON = 0.0001;
+    static final Random r = new Random();
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/util/FastInvTrigTest.java:32]: we don't use this random, use the one provided by the test framework so that seeds are reproducible.

**Review by madrob — COMMENTED** [review]
This looks interesting, and I don't quite understand the math behind all of it, but I left some comments about the code itself.

File: solr/core/src/java/org/apache/solr/util/FastInvTrig.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.util;
+
+public class FastInvTrig {
```
[the remark this answers is not in the document]
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/util/FastInvTrig.java:21]: Yes I wrote this, original in my github repo. It's a Maclaurin series expansion of acos. TABLE stores the coefficients that can be re-used for subsequent calculations, it also reuses x^2 for x^3,x^5 etc. Also initially my implementation required a lot of terms for convergence, until I found this https://stackoverflow.com/questions/20196000/own-asin-function-with-taylor-series-not-accurate which allows for faster convergence near -1,1. I have a benchmark in my repo to show it more performant than Math.acos or FastMath.acos, accuracy appears OK in my tests. NVector with acos faster than Haversine.

File: solr/core/src/test/org/apache/solr/util/FastInvTrigTest.java (hunk 5 above)
Comment (danrosher) [reply solr/core/src/test/org/apache/solr/util/FastInvTrigTest.java:58]: changed to assertEquals

File: solr/core/src/java/org/apache/solr/search/function/distance/NVector.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+import org.apache.lucene.queries.function.valuesource.MultiValueSource;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.solr.common.SolrException;
+
+import java.io.IOException;
+import java.util.Map;
+
+import static org.apache.solr.util.NVectorUtil.NVectorDist;
+
+public class NVector extends ValueSource {
```
[the remark this answers is not in the document]
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/search/function/distance/NVector.java]: changed now

File: solr/core/src/test/org/apache/solr/util/NVectorUtilTest.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    public void latLongToNVector() {
+        double lat = 52.024535;
+        double lon = -0.490155;
+        double[] n = NVectorUtil.latLongToNVector(lat, lon);
+        double[] ll = NVectorUtil.NVectorToLatLong(n);
+        assertSimilar(lat, ll[0]);
+        assertSimilar(lon, ll[1]);
+    }
+
+    void assertSimilar(double expected, double actual) {
```
[the remark this answers is not in the document]
Comment (danrosher) [reply solr/core/src/test/org/apache/solr/util/NVectorUtilTest.java]: changed to `assertEquals(double, double, delta)`

File: solr/core/src/java/org/apache/solr/util/NVectorUtil.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.util;
+
+public class NVectorUtil {
+
+    public static final double EARTH_RADIUS = 6173.008;//km google:standard mean earth radius;
```
[the remark this answers is not in the document]
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/util/NVectorUtil.java]: Oops! Yes using org.locationtech.spatial4j.distance.DistanceUtils.EARTH_MEAN_RADIUS_KM now

File: solr/core/src/test-files/solr/collection1/conf/schema-nvector.xml (hunk 4 above)
Comment (danrosher) [reply solr/core/src/test-files/solr/collection1/conf/schema-nvector.xml:28]: reduced schema size

File: solr/core/src/test/org/apache/solr/util/FastInvTrigTest.java (hunk 6 above)
Comment (danrosher) [reply solr/core/src/test/org/apache/solr/util/FastInvTrigTest.java]: using random() now

File: solr/core/src/java/org/apache/solr/schema/NVector.java (hunk 1 above)
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/schema/NVector.java]: I've added a separator init param and parse doubles using NumberFormat with default locale

**Comment by madrob** [comment]
I found your source `FastInvTrig` repo and ran the benchmarks from there, with a few tweaks to improve measurement accuracy and also added Lucene's [SloppyMath](https://javadoc.io/doc/org.apache.lucene/lucene-core/latest/org/apache/lucene/util/SloppyMath.html) into the competition.

My results are directionally the same as yours - 

```
Benchmark                                    Mode  Cnt    Score    Error  Units
FastInvTrigBenchmark.acosBM                  avgt    3   17.576 ±  1.243  ns/op
FastInvTrigBenchmark.fastMathAcosBM          avgt    3   87.301 ±  4.303  ns/op
FastInvTrigBenchmark.haversineBM             avgt    3   95.062 ± 17.137  ns/op
FastInvTrigBenchmark.mathAcosBM              avgt    3  140.391 ±  5.508  ns/op
FastInvTrigBenchmark.sloppyHaversineMeters   avgt    3   46.647 ±  8.927  ns/op
FastInvTrigBenchmark.sloppyHaversineSortKey  avgt    3   33.637 ±  6.364  ns/op
```

One concern for comparison that I have would be that the current Lucene implementation (which is only 2x slower than yours) has an upper error bound of 40cm. I think we can get there with your series expansions by adding additional terms, but I don't remember my calculus well enough to know how many we need. If we do that, how much does performance suffer? Are we still competitive? You're currently at 10m, which I suspect might be too large of a delta to be useful for the applications that I am familiar with.

**Comment by madrob** [comment]
I'm a little confused as to what the sloppyAsin results are...

From your tests, we should still prefer NVector with MacLaurian expansion at 17 terms over using Sloppy Haversine, right? Is Sloppy Asin using NVector with no series expansion for the trig and instead the lookup tables from Sloppy Math? Which gives us the 10cm accuracy at _almost_ the original 10 term expansion performance level?

**Comment by madrob** [comment]
Can we do a similar trick to split the calculation to get an n-vector sort key and an n-vector meters and get even more speedup for the cases where we don't care about absolute distances?

**Comment by danrosher** [comment]
> I'm a little confused as to what the sloppyAsin results are...
> 
> From your tests, we should still prefer NVector with MacLaurian expansion at 17 terms over using Sloppy Haversine, right? Is Sloppy Asin using NVector with no series expansion for the trig and instead the lookup tables from Sloppy Math? Which gives us the 40cm accuracy at _almost_ the original 10 term expansion performance level?

We calculate the Great circle distance as `d=R*acos(a.b)` where d = distance, R = radius, and a,b are NVectors (a.b is the scalar dot product)

We also know `acos(x) = pi/2-asin(x)`

So I compared FastMath.acos (with 17 terms) with SloppyMath.asin, and found that FastMath.acos with 17 terms is 1.3x slower than SloppyMath.asin.

This is what I meant with FastInvTrigBenchmark.sloppyAsin

so SloppyMath.asin, at the required precision, is faster than FastMath.acos, 

So I was thinking of abandoning FastMath.acos in favour of SloppyMath.asin, what do you think?

> Can we do a similar trick to split the calculation to get an n-vector sort key and an n-vector meters and get even more speedup for the cases where we don't care about absolute distances?

Yes! From looking at the acos plot ( https://www.wolframalpha.com/input?i2d=true&i=acos%5C%2840%29x%5C%2841%29 ) it's a 1 to 1 function, so well suited for comparison. I did a quick test which confirmed that the dot product is enough for comparison between values (which is all we need to do, as we cache the nvectors in the NVectorField). So we then have the following benchmark (with NVectorSortKey as this comparison and the fastest) :

```
Benchmark                                   (num_points)  Mode  Cnt    Score   Error  Units
FastInvTrigBenchmark.NVectorSortKey              2000000  avgt    2   13.846          ms/op
FastInvTrigBenchmark.acosBM10                    2000000  avgt    2   45.620          ms/op
FastInvTrigBenchmark.acosBM17                    2000000  avgt    2   68.673          ms/op
FastInvTrigBenchmark.fastMathAcosBM              2000000  avgt    2  249.075          ms/op
FastInvTrigBenchmark.haversineBM                 2000000  avgt    2  289.814          ms/op
FastInvTrigBenchmark.mathAcosBM                  2000000  avgt    2  304.411          ms/op
FastInvTrigBenchmark.mathAsin                    2000000  avgt    2  323.126          ms/op
FastInvTrigBenchmark.sloppyAsin                  2000000  avgt    2   49.749          ms/op
FastInvTrigBenchmark.sloppyHaversinMeters        2000000  avgt    2  135.847          ms/op
FastInvTrigBenchmark.sloppyHaversinSortKey       2000000  avgt    2   82.149          ms/op
```

So in Solr perhaps we can use NVectorSortKey for sort comparisons then.

**Comment by madrob** [comment]
Ok, I get it now. Yes, let's do the SloppyMath.asin approach, splitting out the sort key (I would call it dot product and add comments about why we can use it as sort key instead of calling the method `SortKey`)

**Comment by danrosher** [comment]
- SloppyMath.asin now replaces FastInvTrig.acos
- Sorting in the function query is on the dot product, values still fetch distance

File: solr/core/src/java/org/apache/solr/util/NVectorUtil.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+  public static double NVectorDotProduct(double[] a, double[] b) {
+    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
+  }
+
+  public static double NVectorDist(double[] a, double[] b) {
+    return NVectorDist(a, b, EARTH_MEAN_RADIUS_KM);
+  }
+
+  public static double NVectorDist(double[] a, double[] b, double radius) {
+    return radius * (pip2 - SloppyMath.asin(a[0] * b[0] + a[1] * b[1] + a[2] * b[2]));
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/util/NVectorUtil.java:80]:
```suggestion
    return NVectorDist(NVectorDotProduct(a, b), radius);
```

File: solr/core/src/java/org/apache/solr/schema/NVectorField.java [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+     * Given a string containing <i>dimension</i> values encoded in it, separated by commas, return a
+     * String array of length <i>dimension</i> containing the values.
+     *
+     * @param externalVal The value to parse
+     * @param dimension   The expected number of values for the point
+     * @param separator   The separator between values
+     * @return An array of the values that make up the point (aka vector)
+     * @throws SolrException if the dimension specified does not match the number found
+     */
+    public static String[] parseCommaSeparatedList(String externalVal, int dimension, String separator)
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/schema/NVectorField.java:103]: Can we have some unit tests for this? Parsing is one of those thing that seems easy but there are always edge cases and it's easy to introduce regressions.

**Review by madrob — COMMENTED** [review]
I think the core of the idea is good here. Big thing missing is an update to the ref guide, probably function-queries.adoc or spatial-search.adoc?

Another thought I had was how exactly do we expect users to use this. If they're still going to be providing indexable data in lat/long and also expecting lat/long for output information, then will this really be faster than using haversine? Or does it move the computation to the indexing side when we only have to do it once, so over multiple queries the total time taken gets reduced...

File: solr/core/src/java/org/apache/solr/schema/NVectorField.java [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+
+import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+public class NVectorField extends CoordinateFieldType {
```
[the remark this answers is not in the document]
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/schema/NVectorField.java:41]: NVectorField is different I think to PointType. NVector replaces lat/lon and overrides most of the important methods in PointType anyway. NVectorField would need a specialized getSpecializedRangeQuery too (not sure how to implement this for nvector yet)

File: solr/core/src/java/org/apache/solr/schema/NVectorField.java [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+
+    String DEFAULT_SEPARATOR = ",";
```
[the remark this answers is not in the document]
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/schema/NVectorField.java]: updated

File: solr/core/src/java/org/apache/solr/util/NVectorUtil.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+import static org.locationtech.spatial4j.distance.DistanceUtils.EARTH_MEAN_RADIUS_KM;
+
+public class NVectorUtil {
+
+  private static final double pip2 = Math.PI/2;
+
+  public static double[] latLongToNVector(double lat, double lon) {
+    double latRad = lat * (Math.PI / 180);
+    double lonRad = lon * (Math.PI / 180);
```
[the remark this answers is not in the document]
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/util/NVectorUtil.java]: updated

File: solr/core/src/java/org/apache/solr/util/NVectorUtil.java (hunk 11 above)
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/util/NVectorUtil.java]: thanks for suggestion, updated

**Comment by danrosher** [comment]
> I think the core of the idea is good here. Big thing missing is an update to the ref guide, probably function-queries.adoc or spatial-search.adoc?
> 

When i get a moment I'll ad something to spatial-search.adoc perhaps?

> Another thought I had was how exactly do we expect users to use this. If they're still going to be providing indexable data in lat/long and also expecting lat/long for output information, then will this really be faster than using haversine? Or does it move the computation to the indexing side when we only have to do it once, so over multiple queries the total time taken gets reduced...

This moves most of the calculation to the indexing side. We then only need to calculate an n-vector for the input lat/lon. The sorting can then be done on the dot-product alone. Users can optionally index spatial data in multiple formats (e.g. LanLonSpatialField and NVectorField) should they find that a performance boost. N-Vector provides faster comparison  (and great circle distance calculation) than haversine, additionally without caveats, or accuracy degradation, for calculations at poles/equator etc.

File: solr/core/src/java/org/apache/solr/util/NVectorUtil.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  public static String[] latLongToNVector(String[] point, NumberFormat formatter) throws ParseException {
+    double[] nvec = latLongToNVector(formatter.parse(point[0]).doubleValue(),formatter.parse(point[1]).doubleValue());
+    return new String[] {
+            Double.toString(nvec[0]), Double.toString(nvec[1]), Double.toString(nvec[2])
+    };
+  }
+
+  public static double[] NVectorToLatLong(double[] n) {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/util/NVectorUtil.java:58]: nit: method names should start with lowercase (here and others)

File: solr/core/src/java/org/apache/solr/util/NVectorUtil.java [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+package org.apache.solr.util;
+
+import org.apache.lucene.util.SloppyMath;
+
+import java.text.NumberFormat;
+import java.text.ParseException;
+
+import static org.locationtech.spatial4j.distance.DistanceUtils.EARTH_MEAN_RADIUS_KM;
+
+public class NVectorUtil {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/util/NVectorUtil.java:27]: can we add javadoc for the class and methods?

File: solr/core/src/java/org/apache/solr/util/NVectorUtil.java [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+  public static String[] latLongToNVector(String[] point, NumberFormat formatter) throws ParseException {
+    double[] nvec = latLongToNVector(formatter.parse(point[0]).doubleValue(),formatter.parse(point[1]).doubleValue());
+    return new String[] {
+            Double.toString(nvec[0]), Double.toString(nvec[1]), Double.toString(nvec[2])
+    };
+  }
+
+  public static double[] NVectorToLatLong(double[] n) {
+    return new double[] {
+      Math.asin(n[2]) * (180 / Math.PI), Math.atan(n[1] / n[0]) * (180 / Math.PI)
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/util/NVectorUtil.java:60]: Math.toDegrees

File: solr/core/src/test/org/apache/solr/search/function/distance/NVectorDistTest.java [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+            + "'nvector_0_d1':0.6152990562577377,"
+            + "'nvector_1_d1':-0.005263047078845837,"
+            + "'nvector_2_d1':0.7882762026750415,"
+            + "'nvector':'52.02471051274793, -0.49007556238612354'}");
+
+    assertJQ(
+        req(
+            "defType", "lucene",
+            "q", "*:*",
+            "nvd", "nvdist(52.01966071979866, -0.4983083573742952,nvector)",
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/search/function/distance/NVectorDistTest.java:59]: I missed where 'nvdist' is defined as a function. Do we need to add to `ValueSourceParser` static init block?

**Review by madrob — COMMENTED** [review]
Few more comments, mainly about documentation of the feature.

File: solr/core/src/java/org/apache/solr/util/NVectorUtil.java (hunk 16 above)
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/util/NVectorUtil.java]: updated

File: solr/core/src/java/org/apache/solr/util/NVectorUtil.java (hunk 17 above)
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/util/NVectorUtil.java:34]: added

File: solr/core/src/java/org/apache/solr/util/NVectorUtil.java (hunk 18 above)
Comment (danrosher) [reply solr/core/src/java/org/apache/solr/util/NVectorUtil.java]: changed

File: solr/core/src/test/org/apache/solr/search/function/distance/NVectorDistTest.java (hunk 19 above)
Comment (danrosher) [reply solr/core/src/test/org/apache/solr/search/function/distance/NVectorDistTest.java:59]: added to ValueSourceParser
